File: lect-player-web/src/main/frontend/src/tool/select-group.tool.ts
import { Tool } from "./tool";
import { ToolContext } from "./tool-context";
import { PenPoint } from "../geometry/pen-point";
import { Shape } from "../model/shape/shape";
import { SelectShape } from "../model/shape/select.shape";
import { Rectangle } from "../geometry/rectangle";

enum Mode { Select, Move }

class SelectGroupTool implements Tool {

	private mode: Mode;
	
	private context: ToolContext;
	
	private shape: SelectShape;
	
	private selectedShapes: Shape[];
	
	private sourcePoint: PenPoint;
	
	private initialized: boolean;


	begin(point: PenPoint, context: ToolContext): void {
		this.sourcePoint = point.clone();
		this.context = context;
		this.initialized = false;

		this.shape = new SelectShape();

		this.context.page.addShape(this.shape);

		this.getSelectedShapes();

		if (this.hasSelectedShapes()) {
			if (this.hitSelected(point)) {
				this.mode = Mode.Move;
			}
			else {
				this.context.beginBulkRender();
				this.removeSelection();
				this.context.endBulkRender();

				this.mode = Mode.Select;
			}
		}
		else {
			this.mode = Mode.Select;
		}
	}

	execute(point: PenPoint): void {
		if (this.mode == Mode.Select) {
			if (!this.initialized) {
				this.shape.setP0(point);
				this.initialized = true;
			}

			this.shape.addPoint(point.clone());
			this.shape.setP1(point);

			this.selectGroup(this.shape.bounds);
		}
		else if (this.mode == Mode.Move) {
			this.sourcePoint.subtract(point);

			this.moveShapes(this.sourcePoint);

			this.sourcePoint = point.clone();
		}
	}

	end(point: PenPoint): void {
		this.context.page.removeShape(this.shape);

		this.initialized = false;
	}

	selectGroup(rect: Rectangle): void {
		this.context.beginBulkRender();

		this.removeSelection();

		for (let shape of this.context.page.getShapes()) {
			if (shape === this.shape) {
				continue;
			}
			if (shape.intersects(rect)) {
				this.addSelection(shape);
			}
		}

		this.context.endBulkRender();
	}

	private addSelection(shape: Shape): void {
		shape.setSelected(true);

		this.selectedShapes.push(shape);
	}

	private getSelectedShapes(): void {
		this.selectedShapes = new Array<Shape>();

		for (let shape of this.context.page.getShapes()) {
			if (shape.isSelected()) {
				this.selectedShapes.push(shape);
			}
		}
	}

	private hasSelectedShapes(): boolean {
		return this.selectedShapes.length != 0;
	}

	private hitSelected(point: PenPoint): boolean {
		for (let shape of this.selectedShapes) {
			if (shape.contains(point)) {
				return true;
			}
		}

		return false;
	}

	private removeSelection(): void {
		for (let shape of this.selectedShapes) {
			shape.setSelected(false);
		}

		this.selectedShapes.length = 0;
	}

	private moveShapes(delta: PenPoint): void {
		if (!this.hasSelectedShapes()) {
			return;
		}

		this.context.beginBulkRender();

		for (let shape of this.selectedShapes) {
			shape.moveByDelta(delta);
		}

		this.context.endBulkRender();
	}
}

export { SelectGroupTool };

File: lect-core/src/main/java/org/lecturestudio/core/tool/SelectGroupTool.java
/*
 * Copyright (C) 2020 TU Darmstadt, Department of Computer Science,
 * Embedded Systems and Applications Group.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.lecturestudio.core.tool;

import java.util.ArrayList;
import java.util.List;

import org.lecturestudio.core.geometry.PenPoint2D;
import org.lecturestudio.core.geometry.Rectangle2D;
import org.lecturestudio.core.model.Page;
import org.lecturestudio.core.model.shape.SelectShape;
import org.lecturestudio.core.model.shape.Shape;
import org.lecturestudio.core.recording.action.SelectGroupAction;

public class SelectGroupTool extends Tool {

	private enum Mode {
		Select,
		Move
	}



	private Mode mode;

	private Page page;

	private SelectShape shape;

	private List<Shape> selectedShapes;

	private PenPoint2D sourcePoint;


	public SelectGroupTool(ToolContext context) {
		super(context);
	}

	@Override
	public void begin(PenPoint2D point, Page page) {
		recordAction(new SelectGroupAction());

		super.begin(point, page);

		this.page = page;
		this.sourcePoint = point.clone();

		shape = new SelectShape(createStroke());
		shape.setStartPoint(point.clone());

		page.addShape(shape);

		fireToolEvent(new ShapePaintEvent(ToolEventType.BEGIN, shape, shape.getBounds()));

		getSelectedShapes(page);

		if (hasSelectedShapes()) {
			if (hitSelected(point)) {
				mode = Mode.Move;
			}
			else {
				removeSelection();

				mode = Mode.Select;
			}
		}
		else {
			mode = Mode.Select;
		}
	}

	@Override
	public void execute(PenPoint2D point) {
		if (mode == Mode.Select) {
			Rectangle2D dirtyArea = shape.getBounds().clone();

			shape.setEndPoint(point);

			dirtyArea.union(shape.getBounds());

			fireToolEvent(new ShapePaintEvent(ToolEventType.EXECUTE, shape, dirtyArea));

			selectGroup(shape.getBounds());
		}
		else if (mode == Mode.Move) {
			sourcePoint.subtract(point);

			moveShapes(sourcePoint);

			sourcePoint.set(point);
		}

		super.execute(point);
	}

	@Override
	public void end(PenPoint2D point) {
		page.removeShape(shape);

		super.end(point);

		fireToolEvent(new ShapePaintEvent(ToolEventType.END, shape, shape.getBounds()));
	}

	@Override
	public ToolType getType() {
		return ToolType.SELECT_GROUP;
	}

	private void getSelectedShapes(Page page) {
		selectedShapes = new ArrayList<>();

		for (Shape shape : page.getShapes()) {
			if (shape.isSelected()) {
				selectedShapes.add(shape);
			}
		}
	}

	private void moveShapes(PenPoint2D delta) {
		for (Shape shape : selectedShapes) {
			shape.moveByDelta(delta);
		}

		if (!selectedShapes.isEmpty()) {
			fireToolEvent(new ShapeModifyEvent(selectedShapes));
		}
	}

	private boolean hasSelectedShapes() {
		return !selectedShapes.isEmpty();
	}

	private void addSelection(Shape shape) {
		shape.setSelected(true);

		selectedShapes.add(shape);
	}

	private void selectGroup(Rectangle2D rect) {
		removeSelection();

		for (Shape shape : page.getShapes()) {
			if (shape.intersects(rect)) {
				addSelection(shape);
			}
		}

		if (!selectedShapes.isEmpty()) {
			fireToolEvent(new ShapeModifyEvent(selectedShapes));
		}
	}

	private void removeSelection() {
		for (Shape shape : selectedShapes) {
			shape.setSelected(false);
		}

		selectedShapes.clear();
	}

	private boolean hitSelected(PenPoint2D point) {
		for (Shape shape : selectedShapes) {
			if (shape.contains(point)) {
				return true;
			}
		}

		return false;
	}

	private Stroke createStroke() {
		StrokeSettings settings = context.getPaintSettings(ToolType.RECTANGLE);

		Stroke stroke = new Stroke();
		stroke.setColor(settings.getColor().derive(settings.getAlpha()));
		stroke.setWidth(settings.getWidth());

		return stroke;
	}
}
